File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/otaserver/LocalOtaServer.kt
package com.mentra.bluetoothsdk.otaserver

import com.mentra.bluetoothsdk.debug.BleTraceLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.BufferedInputStream
import java.io.ByteArrayOutputStream
import java.io.EOFException
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.nio.charset.StandardCharsets
import java.util.Locale
import java.util.concurrent.Semaphore
import org.json.JSONObject

/**
 * Static HTTP server for hotspot-served glasses OTA (OS-1676).
 *
 * Serves the rewritten OTA manifest at /version.json and pre-downloaded artifacts at
 * /artifacts/<sha256> to the glasses over the glasses-hotspot link. Same hand-rolled
 * socket structure as [LocalPhotoUploadServer] —
 * no third-party HTTP dependency in the SDK.
 */
class LocalOtaServer(
    private val onLog: (String) -> Unit,
) : AutoCloseable {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var serverSocket: ServerSocket? = null
    private var serverJob: Job? = null
    private val clientSlots = Semaphore(MAX_ACTIVE_CLIENTS)

    @Volatile
    private var manifestJson: String = "{}"

    @Volatile
    private var artifacts: Map<String, File> = emptyMap()

    @Volatile
    var running: Boolean = false
        private set

    val activePort: Int?
        get() = serverSocket
            ?.takeIf { running && !it.isClosed }
            ?.localPort

    /** Swap the served manifest and artifact set. Safe while the server is running. */
    fun configure(manifestJson: String, artifacts: Map<String, File>) {
        this.manifestJson = manifestJson
        this.artifacts = artifacts
    }

    fun start(host: String, port: Int): Int {
        val activeSocket = serverSocket
        if (running && activeSocket != null && !activeSocket.isClosed && activeSocket.localPort == port) {
            onLog("Already listening on $host:${activeSocket.localPort}")
            return activeSocket.localPort
        }
        stop()
        val socket = ServerSocket()
        try {
            socket.reuseAddress = true
            socket.bind(InetSocketAddress(host, port))
            serverSocket = socket
            running = true
            serverJob = scope.launch {
                acceptLoop(socket)
            }
        } catch (error: Throwable) {
            try {
                socket.close()
            } catch (_: Throwable) {
            }
            throw error
        }
        onLog("OTA server listening on $host:${socket.localPort}")
        return socket.localPort
    }

    fun stop() {
        running = false
        serverJob?.cancel()
        serverJob = null
        try {
            serverSocket?.close()
        } catch (_: Throwable) {
        }
        serverSocket = null
    }

    override fun close() {
        stop()
        scope.cancel()
    }

    private fun acceptLoop(socket: ServerSocket) {
        while (scope.isActive && !socket.isClosed) {
            try {
                val client = socket.accept()
                if (!clientSlots.tryAcquire()) {
                    rejectClient(client)
                    continue
                }
                scope.launch {
                    try {
                        handleClient(client)
                    } finally {
                        clientSlots.release()
                    }
                }
            } catch (error: SocketException) {
                if (running) onLog("Accept failed: ${error.message}")
            } catch (error: Throwable) {
                if (running) onLog("Accept failed: ${error.message ?: error::class.java.simpleName}")
            }
        }
    }

    private fun handleClient(socket: Socket) {
        socket.use { client ->
            val requestStartMs = System.currentTimeMillis()
            client.soTimeout = SOCKET_TIMEOUT_MS
            val input = BufferedInputStream(client.getInputStream())
            val output = client.getOutputStream()
            var request: HttpRequest? = null
            var status = 0
            var responseBytes = 0L

            try {
                val headerText = readHeaders(input).toString(StandardCharsets.ISO_8859_1)
                request = HttpRequest.parse(headerText)
                onLog("${request.method} ${request.path} from ${client.inetAddress.hostAddress}")

                if (request.method != "GET") {
                    status = 405
                    writeJson(output, status, """{"ok":false,"error":"method_not_allowed"}""")
                    return
                }

                when {
                    request.path == "/version.json" -> {
                        val body = manifestJson.toByteArray(StandardCharsets.UTF_8)
                        status = 200
                        responseBytes = body.size.toLong()
                        writeBody(output, status, "application/json", body)
                    }
                    request.path.startsWith("/artifacts/") -> {
                        val key = request.path.removePrefix("/artifacts/")
                        val file = artifacts[key]
                        if (file == null || !file.isFile) {
                            status = 404
                            writeJson(output, status, """{"ok":false,"error":"artifact_not_found"}""")
                        } else {
                            status = 200
                            responseBytes = file.length()
                            writeFile(output, file)
                        }
                    }
                    else -> {
                        status = 404
                        writeJson(output, status, """{"ok":false,"error":"not_found"}""")
                    }
                }
            } catch (error: Throwable) {
                onLog("request failed: ${error.message ?: error::class.java.simpleName}")
                if (status == 0) {
                    status = 500
                    try {
                        writeJson(output, status, """{"ok":false,"error":"server_error"}""")
                    } catch (_: Throwable) {
                    }
                }
            } finally {
                trace(client, request, status, responseBytes, requestStartMs)
            }
        }
    }

    private fun writeFile(output: OutputStream, file: File) {
        val fileLength = file.length()
        output.write("HTTP/1.1 200 OK\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Content-Type: application/octet-stream\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Content-Length: $fileLength\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Connection: close\r\n\r\n".toByteArray(StandardCharsets.US_ASCII))
        file.inputStream().use { fileInput ->
            val buffer = ByteArray(STREAM_CHUNK_BYTES)
            var remaining = fileLength
            while (remaining > 0) {
                val read = fileInput.read(buffer, 0, minOf(buffer.size.toLong(), remaining).toInt())
                if (read <= 0) throw EOFException("artifact truncated while streaming")
                output.write(buffer, 0, read)
                remaining -= read
            }
        }
        output.flush()
    }

    private fun rejectClient(socket: Socket) {
        socket.use { client ->
            try {
                writeJson(client.getOutputStream(), 503, """{"ok":false,"error":"too_many_connections"}""")
            } catch (_: Throwable) {
            }
        }
    }

    private fun readHeaders(input: InputStream): ByteArray {
        val out = ByteArrayOutputStream()
        val delimiter = byteArrayOf(13, 10, 13, 10)
        var matched = 0
        while (true) {
            val byte = input.read()
            if (byte == -1) throw EOFException("Socket closed before HTTP headers completed")
            out.write(byte)
            matched = if (byte.toByte() == delimiter[matched]) {
                matched + 1
            } else if (byte.toByte() == delimiter[0]) {
                1
            } else {
                0
            }
            if (matched == delimiter.size) return out.toByteArray()
            if (out.size() > MAX_HEADER_BYTES) throw IllegalArgumentException("HTTP headers too large")
        }
    }

    private fun writeJson(output: OutputStream, status: Int, body: String) {
        writeBody(output, status, "application/json", body.toByteArray(StandardCharsets.UTF_8))
    }

    private fun writeBody(
        output: OutputStream,
        status: Int,
        contentType: String,
        body: ByteArray,
    ) {
        val reason = when (status) {
            200 -> "OK"
            404 -> "Not Found"
            405 -> "Method Not Allowed"
            503 -> "Service Unavailable"
            else -> "Internal Server Error"
        }
        output.write("HTTP/1.1 $status $reason\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Content-Type: $contentType\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Content-Length: ${body.size}\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write("Connection: close\r\n\r\n".toByteArray(StandardCharsets.US_ASCII))
        output.write(body)
        output.flush()
    }

    private fun trace(
        client: Socket,
        request: HttpRequest?,
        status: Int,
        responseBytes: Long,
        startMs: Long,
    ) {
        val payload = JSONObject()
        try {
            payload.put("type", "ota_server_request")
            payload.put("remoteHost", client.inetAddress?.hostAddress ?: "")
            payload.put("method", request?.method ?: "")
            payload.put("path", request?.path ?: "")
            payload.put("statusCode", status)
            payload.put("responseBytes", responseBytes)
            payload.put("durationMs", System.currentTimeMillis() - startMs)
            BleTraceLogger.logJson("phone_to_wifi", "wifi_http_output", payload, null)
        } catch (_: Throwable) {
        }
    }

    private data class HttpRequest(
        val method: String,
        val path: String,
    ) {
        companion object {
            fun parse(headerText: String): HttpRequest {
                val lines = headerText.split("\r\n").filter { it.isNotBlank() }
                val requestParts = lines.firstOrNull()?.split(" ").orEmpty()
                val method = requestParts.getOrNull(0)?.uppercase(Locale.US).orEmpty()
                val path = requestParts.getOrNull(1)?.substringBefore("?").orEmpty()
                return HttpRequest(method, path)
            }
        }
    }

    companion object {
        private const val MAX_HEADER_BYTES = 64 * 1024
        private const val MAX_ACTIVE_CLIENTS = 2
        private const val SOCKET_TIMEOUT_MS = 20_000
        private const val STREAM_CHUNK_BYTES = 64 * 1024
    }
}

File: mobile/modules/bluetooth-sdk/ios/LocalOtaServer.swift
import Foundation
import Network

struct OtaServerError: LocalizedError {
  let message: String

  init(_ message: String) {
    self.message = message
  }

  var errorDescription: String? { message }
}

/// Static HTTP server for hotspot-served glasses OTA (OS-1676).
///
/// Serves the rewritten OTA manifest at /version.json and pre-downloaded artifacts at
/// /artifacts/<sha256> to the glasses over the glasses-hotspot link. Mirrors
/// LocalPhotoUploadServer's NWListener structure.
final class LocalOtaServer {
  private let queue = DispatchQueue(label: "com.mentra.ota-server")
  private let onLog: (String) -> Void
  private var listener: NWListener?
  private var listenerGeneration = 0
  private(set) var activePort: UInt16?

  private var manifestJson = "{}"
  private var artifacts: [String: URL] = [:]

  private static let maxHeaderBytes = 64 * 1024
  private static let streamChunkBytes = 256 * 1024

  init(onLog: @escaping (String) -> Void) {
    self.onLog = onLog
  }

  /// Swap the served manifest and artifact set. Safe while the server is running.
  func configure(manifestJson: String, artifacts: [String: URL]) {
    queue.sync {
      self.manifestJson = manifestJson
      self.artifacts = artifacts
    }
  }

  func start(host: String, port: UInt16) throws -> UInt16 {
    if listener != nil, activePort == port {
      onLog("Already listening on \(host):\(port)")
      return port
    }
    stop()

    let parameters = NWParameters.tcp
    parameters.allowLocalEndpointReuse = true
    guard let endpointPort = NWEndpoint.Port(rawValue: port) else {
      throw OtaServerError("Invalid port \(port)")
    }

    parameters.requiredLocalEndpoint = .hostPort(host: NWEndpoint.Host(host), port: endpointPort)
    let listener = try NWListener(using: parameters)
    listenerGeneration += 1
    let generation = listenerGeneration
    let started = DispatchSemaphore(value: 0)
    var startError: Error?
    var isReady = false

    listener.stateUpdateHandler = { [weak self] state in
      guard let self, self.listenerGeneration == generation else {
        return
      }
      switch state {
      case .ready:
        isReady = true
        self.activePort = port
        self.onLog("OTA server listening on \(host):\(port)")
        started.signal()
      case .failed(let error):
        startError = error
        self.activePort = nil
        self.listener = nil
        self.onLog("OTA server failed on \(port): \(error)")
        started.signal()
      case .cancelled:
        self.activePort = nil
        if !isReady {
          startError = OtaServerError("Listener cancelled")
          started.signal()
        }
      default:
        break
      }
    }

    listener.newConnectionHandler = { [weak self] connection in
      self?.handle(connection)
    }

    self.listener = listener
    listener.start(queue: queue)

    if started.wait(timeout: .now() + 2) == .timedOut {
      self.listener = nil
      listener.cancel()
      throw OtaServerError("Timed out starting listener")
    }
    if let startError {
      self.listener = nil
      listener.cancel()
      throw startError
    }

    return port
  }

  func stop() {
    listenerGeneration += 1
    listener?.cancel()
    listener = nil
    activePort = nil
  }

  private final class RequestReadState {
    var buffer = Data()
  }

  private func handle(_ connection: NWConnection) {
    connection.start(queue: queue)
    receive(connection, state: RequestReadState())
  }

  private func receive(_ connection: NWConnection, state: RequestReadState) {
    connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) {
      [weak self] data, _, isComplete, error in
      guard let self else {
        connection.cancel()
        return
      }
      if let error {
        self.onLog("request failed: \(error.localizedDescription)")
        connection.cancel()
        return
      }
      if let data, !data.isEmpty {
        state.buffer.append(data)
      }

      if let headerRange = state.buffer.range(of: Data([13, 10, 13, 10])) {
        let headerData = state.buffer.subdata(in: state.buffer.startIndex..<headerRange.lowerBound)
        let headerText = String(data: headerData, encoding: .isoLatin1) ?? ""
        self.route(connection, headerText: headerText)
        return
      }
      if state.buffer.count > Self.maxHeaderBytes {
        self.writeJson(connection, status: 400, body: #"{"ok":false,"error":"headers_too_large"}"#)
        return
      }
      if isComplete {
        connection.cancel()
        return
      }
      self.receive(connection, state: state)
    }
  }

  private func route(_ connection: NWConnection, headerText: String) {
    let lines = headerText.components(separatedBy: "\r\n").filter { !$0.isEmpty }
    let requestParts = lines.first?.components(separatedBy: " ") ?? []
    let method = requestParts.count > 0 ? requestParts[0].uppercased() : ""
    let rawPath = requestParts.count > 1 ? requestParts[1] : ""
    let path = rawPath.components(separatedBy: "?")[0]
    onLog("\(method) \(path)")

    guard method == "GET" else {
      writeJson(connection, status: 405, body: #"{"ok":false,"error":"method_not_allowed"}"#)
      return
    }

    if path == "/version.json" {
      writeBody(
        connection,
        status: 200,
        contentType: "application/json",
        body: Data(manifestJson.utf8)
      )
      return
    }

    if path.hasPrefix("/artifacts/") {
      let key = String(path.dropFirst("/artifacts/".count))
      guard let fileUrl = artifacts[key],
            let fileSize = try? fileUrl.resourceValues(forKeys: [.fileSizeKey]).fileSize
      else {
        writeJson(connection, status: 404, body: #"{"ok":false,"error":"artifact_not_found"}"#)
        return
      }
      sendFile(connection, fileUrl: fileUrl, fileLength: Int64(fileSize))
      return
    }

    writeJson(connection, status: 404, body: #"{"ok":false,"error":"not_found"}"#)
  }

  private func sendFile(
    _ connection: NWConnection,
    fileUrl: URL,
    fileLength: Int64
  ) {
    let header =
      "HTTP/1.1 200 OK\r\n" +
      "Content-Type: application/octet-stream\r\n" +
      "Content-Length: \(fileLength)\r\n" +
      "Connection: close\r\n\r\n"

    guard let fileHandle = try? FileHandle(forReadingFrom: fileUrl) else {
      writeJson(connection, status: 500, body: #"{"ok":false,"error":"artifact_unreadable"}"#)
      return
    }

    connection.send(content: Data(header.utf8), completion: .contentProcessed { [weak self] error in
      guard let self, error == nil else {
        try? fileHandle.close()
        connection.cancel()
        return
      }
      self.streamChunks(connection, fileHandle: fileHandle, remaining: fileLength)
    })
  }

  private func streamChunks(_ connection: NWConnection, fileHandle: FileHandle, remaining: Int64) {
    if remaining <= 0 {
      try? fileHandle.close()
      connection.send(content: nil, contentContext: .finalMessage, isComplete: true, completion: .contentProcessed { _ in
        connection.cancel()
      })
      return
    }
    let chunkSize = Int(min(Int64(Self.streamChunkBytes), remaining))
    let chunk: Data
    do {
      guard let data = try fileHandle.read(upToCount: chunkSize), !data.isEmpty else {
        throw OtaServerError("artifact truncated while streaming")
      }
      chunk = data
    } catch {
      onLog("stream failed: \(error.localizedDescription)")
      try? fileHandle.close()
      connection.cancel()
      return
    }
    connection.send(content: chunk, completion: .contentProcessed { [weak self] error in
      guard let self else {
        try? fileHandle.close()
        connection.cancel()
        return
      }
      if let error {
        self.onLog("stream failed: \(error.localizedDescription)")
        try? fileHandle.close()
        connection.cancel()
        return
      }
      self.streamChunks(connection, fileHandle: fileHandle, remaining: remaining - Int64(chunk.count))
    })
  }

  private func writeJson(
    _ connection: NWConnection,
    status: Int,
    body: String
  ) {
    writeBody(
      connection,
      status: status,
      contentType: "application/json",
      body: Data(body.utf8)
    )
  }

  private func writeBody(
    _ connection: NWConnection,
    status: Int,
    contentType: String,
    body: Data
  ) {
    let reason: String
    switch status {
    case 200:
      reason = "OK"
    case 400:
      reason = "Bad Request"
    case 404:
      reason = "Not Found"
    case 405:
      reason = "Method Not Allowed"
    default:
      reason = "Internal Server Error"
    }

    let header =
      "HTTP/1.1 \(status) \(reason)\r\n" +
      "Content-Type: \(contentType)\r\n" +
      "Content-Length: \(body.count)\r\n" +
      "Connection: close\r\n" +
      "\r\n"
    var response = Data(header.utf8)
    response.append(body)
    connection.send(content: response, completion: .contentProcessed { _ in
      connection.cancel()
    })
  }
}
